lib/core/cmap/connection.js

```diff
@@ -0,0 +1,211 @@
+'use strict';
+
+const EventEmitter = require('events');
+const MessageStream = require('./message_stream');
+const MongoError = require('../error').MongoError;
+const MongoWriteConcernError = require('../error').MongoWriteConcernError;
+const wp = require('../wireprotocol');
+const apm = require('../connection/apm');
+const updateSessionFromResponse = require('../sessions').updateSessionFromResponse;
+
+const kStream = Symbol('stream');
+const kQueue = Symbol('queue');
+const kMessageStream = Symbol('messageStream');
+
+class Connection extends EventEmitter {
+  constructor(stream, options) {
+    super(options);
+
+    this.id = `${stream.address().address}:${stream.address().port}`;
+    this.bson = options.bson;
+    this.description = null;
+    this.socketTimeout = typeof options.socketTimeout === 'number' ? options.socketTimeout : 360000;
+    this.monitorCommands =
+      typeof options.monitorCommands === 'boolean' ? options.monitorCommands : false;
+
+    // setup parser stream and message handling
+    this[kQueue] = new Map();
```

Comment:
I'm assuming this queue will eventually go away, as we will only be allowing one request/response on the socket at a time?